A revamped model of literals suitable for use in Low* code

msprotz opened the pull request:
This is the model and implementation of Low* literals that we chatted about over the past couple weeks. This pull request intentionally does not include LowStar.Printf -- this will come later.

Some things to note:
- one assume, which I believe is inevitable
- a syntactic check is needed on one of the functions, but I'm not sure we can do it using meta-F* (or even want to do it this way, see https://stackoverflow.com/questions/54541577/use-meta-programming-in-f-for-a-syntactic-check-on-a-function-argument)

aseemr commented on ulib/LowStar.Literal.fsti:
```diff
@@ -0,0 +1,137 @@
+module LowStar.Literal
+
+module B = LowStar.Buffer
+module MB = LowStar.Monotonic.Buffer
```
ImmutableBuffer includes LowStar.Monotonic.Buffer. Since you are using only immutable buffers, you could avoid opening these modules, and use for example `IB.recallable`.

aseemr commented on ulib/LowStar.Literal.fsti:
```diff
@@ -0,0 +1,137 @@
+
+val const_string: Type0
+
+val const_string_of_literal: s:string { normalize (zero_free s) } -> const_string
```
How will the clients prove the precondition? `String.list_of_string` is a `val` in `FStar.String.fsti`, and there don't seem to be any lemmas around it?

aseemr commented on ulib/LowStar.Literal.fsti:
```diff
@@ -0,0 +1,137 @@
+    (ensures (fun h0 b h1 ->
+      MB.frameOf b == HS.root /\ // is this really useful?
+      MB.recallable b /\
+      MB.alloc_post_mem_common b h0 h1 (u8s_of_ascii_string s)))
```
This spec will lead to a proof of `False` in the client code. For example:

- Use `B.lemma_equal_mem_domains` to derive that the returned buffer is `live` in `h0` (i.e. in the memory before calling this function).
- So now the buffer is both `unused_in` and `live` in `h0`, leading to proof of false via `B.live_not_unused_in`.

Would it be too bad if it was `ST` function, with `frameOf b == HS.root`?

msprotz replied:
thanks for checking the assumption; I don't mind putting this in ST and I think miTLS folks would be ok with it -- after all, if they're using this to replace allocations of immutable buffers, then this would be the expected signature

msprotz commented:
Pushed a new version of this pull request.
- @aseemr I incorporated your comments; please review the post-condition of buffer_of_literal since it is now assumed (I removed the fst file)
- @nikswamy I removed the requirement on zeroes in the value version
- @nikswamy I made list_ref take a prop (per your original request) but I kept list_refb, otherwise, it would've required my client code to use prop everywhere instead of total functions on lists like List.Tot.for_all, which I find convenient, especially if we later write more computable predicates, such as "the utf8 encoding of this series of scalar values"

Thanks,

Jonathan

msprotz commented:
@nikswamy asks in #1654 whether the const_string type is necessary; under the hood, it will be the same as `Prims.string`; however, the intent is to:
- have `const_string_of_literal` as a marker, to check that its argument is a literal at extraction-time, then
- flag any other subexpression at type `Prims.string` as a non-Low* program (kremlin's warning 15)

This way, other operations on Prims.string (`concat`, etc.) which are currently allowed under the "transitional" regime for miTLS, can be accurately identified and flagged as issues.

nikswamy commented:
I'm still not sure I get the point of the `const_string` type.

According to this interface, the only way to use `const_string` is to print it, as shown below.

```
val const_string: Type0

val const_string_of_literal: s:string -> const_string

val print_string: const_string -> ST.Stack unit
  (requires fun _ -> true)
  (ensures fun h0 _ h1 -> B.(modifies loc_none h0 h1))
```

Wouldn't it be simpler to just have

```
val print_string: string -> ST.Stack unit
  (requires fun _ -> true)
  (ensures fun h0 _ h1 -> B.(modifies loc_none h0 h1))
```

with a KReMLin extraction-time check that ensures that `print_string`'s argument is a literal (complaining about all other uses of strings).

msprotz commented:
This is only the beginning; afterwards, we would have utf8-aware indexing operations (eventually), we could pass such constant strings to printf to use them within a format string, etc.

Even for a debugging routine that takes as an argument an error message, it's useful to be able to pass these around.

msprotz commented:
This pull request is ready for merging:
- strings as values removed, per a discussion with Nik; the `Prims.string` type will continue to be usable in Low* for strings-as-values, with the understanding that any use of a function from FStar.String or of Prims.strcat will trigger a "non-Low*" warning from kremlin
- tweaked a post-condition needed while writing a kremlin test
- added extraction support

msprotz commented:
Updated with regards to master, green on CI.
